src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1052,15 +1053,25 @@ impl LinkCollector<'_, '_> {
             return None;
         }
 
-        let key = ResolutionInfo {
-            module_id,
-            dis: disambiguator,
-            path_str: path_str.to_owned(),
-            extra_fragment,
+        let diag_info = DiagnosticInfo {
+            item,
+            dox,
+            ori_link: &ori_link.link,
+            link_range: ori_link.range.clone(),
         };
-        let diag =
-            DiagnosticInfo { item, dox, ori_link: &ori_link, link_range: link_range.clone() };
-        let (mut res, mut fragment) = self.resolve_with_disambiguator_cached(key, diag)?;
+        let (mut res, mut fragment) = self.resolve_with_disambiguator_cached(
+            ResolutionInfo {
+                module_id,
+                dis: disambiguator,
+                path_str: path_str.to_owned(),
+                extra_fragment,
+                cache_resolution_failure: matches!(
+                    ori_link.kind,
+                    LinkType::Reference | LinkType::Shortcut
```

Comment:
I don't particularly like this, but:
 - LinkType isn't hashable, so it can't be included in the ResolutionInfo struct directly
 - even if LinkType were hashable, including it would result in random test failures (didn't bother investigating)
 - I'm not sure *why* this doesn't break shortcut links completely (maybe because they are parsed as `ShortcutUnknown`?)

Reply:
Calculating whether to cache it ahead of time seems fine. The reason you check explicitly for the kind is because inline links shouldn't be cached (things like `[x](Ok)`)? Can you check for those explicitly instead? Or if that's not what you meant, maybe `Collapsed` should also be checked?

Reply:
And I have another question - I don't think this needs to be part of the key, right? It would work if you passed it as a parameter instead? That would avoid having to hash it at all.

Reply:
It has to be, otherwise "normal" links to the same target as a reference-style link would hit the cache. See the test case for exactly this :) Depends on what you expect, I guess, but this patch aims to emit a single error for each incorrect link definition, or at least that was my goal.

Reply:
Ah, it's also not that links don't get cached, but resolution failures don't get cached. Which is only interesting in the reference style links case, where we don't want to emit an error for each usage of the incorrect link. I think :)

Reply:
> Ah, it's also not that links don't get cached, but resolution failures don't get cached. 

Right, that's what I meant, thanks.

> It has to be, otherwise "normal" links to the same target as a reference-style link would hit the cache.

Ok, I see the issue: if you have something like

```rust
/// [reference][link] and [inline](link)
```
both `link`s should give an error, but if you don't include `cache_resolution_failure` in the cache key, then the second link will pick up the `None` and be returned.

So I have a further question now - when will you hit the cache and *not* have `cache_resolution_failure` set? That would only happen for `[inline](link)` twice in a row, right? But we explicitly choose *not* to insert it into the cache on error. I think there's an optimization here where you can make the size of the cache smaller by explicitly checking `cache_resolution_failure` on a cache hit, that way you don't need that extra bit of info; and the logic seems more clear to me too.

Reply:
> I think there's an optimization here

Interesting thought, that might be something I missed.